File: scripts/initialize_task_loop_state.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Initialize Task Loop State
Initializes project task-loop state files (.agents/task-loop/) with UTF-8 encoding.
Supports auto-scanning project sessions, standardized title generation ([具体专题] 功能点1 & 功能点2),
and managing main_thread_id selection.
"""

import sys
import json
import uuid
import re
import argparse
from pathlib import Path
from typing import List, Dict, Any, Optional

# Ensure UTF-8 stdout
if hasattr(sys.stdout, "reconfigure"):
    try:
        sys.stdout.reconfigure(encoding="utf-8")
    except Exception:
        pass

script_dir = Path(__file__).resolve().parent
if str(script_dir) not in sys.path:
    sys.path.insert(0, str(script_dir))

from find_project_sessions import find_sessions


def generate_standard_title(session: Dict[str, Any]) -> str:
    """
    Generate a standardized title formatted as [具体专题] 功能点1 & 功能点2.
    """
    prompts = " ".join(session.get("recent_prompts", [])).lower()
    touched = " ".join(session.get("recent_touched_files", [])).lower()
    combined = f"{prompts} {touched}"

    # Determine topic category
    if "主会话" in combined or "skill的开发仓库" in combined or "任务编排" in combined:
        category = "主会话"
        func1 = "任务编排"
        func2 = "治理中枢"
    elif "会话" in combined or "session" in combined or "sdk" in combined:
        category = "会话专题"
        func1 = "SDK接口封装"
        func2 = "会话管理"
    elif "memory" in combined or "记忆" in combined:
        category = "记忆专题"
        func1 = "记忆文档维护"
        func2 = "经验沉淀"
    elif "dispatch" in combined or "123" in combined or "complexity" in combined:
        category = "调度专题"
        func1 = "复杂度裁决"
        func2 = "派单协议"
    elif "lease" in combined or "lock" in combined:
        category = "租约专题"
        func1 = "原子锁控制"
        func2 = "并发防护"
    elif "test" in combined or "preflight" in combined:
        category = "测试专题"
        func1 = "门禁验证"
        func2 = "自动化回归"
    else:
        category = "通用专题"
        # Extract first 2 meaningful verbs/nouns from prompt
        raw_prompt = session.get("recent_prompts", [""])[0] if session.get("recent_prompts") else ""
        cleaned = re.sub(r'[^\w\s\u4e00-\u9fa5]', ' ', raw_prompt)
        words = [w for w in cleaned.split() if len(w) >= 2][:2]
        func1 = words[0] if len(words) > 0 else "功能开发"
        func2 = words[1] if len(words) > 1 else "自测验证"

    return f"[{category}] {func1} & {func2}"


def init_state(
    root_path_str: str = ".agents/task-loop",
    project_root_str: str = ".",
    main_thread_id: Optional[str] = None,
    active_vendor: str = "antigravity",
    scan_sessions: bool = True,
    clear_main: bool = False
) -> dict:
    root = Path(root_path_str).resolve()
    root.mkdir(parents=True, exist_ok=True)
    
    sessions_file = root / "sessions.json"
    existing_sessions_data = {}
    if sessions_file.is_file():
        try:
            with open(sessions_file, "r", encoding="utf-8") as f:
                existing_sessions_data = json.load(f)
        except Exception:
            existing_sessions_data = {}

    if clear_main:
        main_thread_id = None
    elif main_thread_id is None:
        main_thread_id = existing_sessions_data.get("main_thread_id")

    # Discover sessions if requested
    discovered_sessions = []
    if scan_sessions:
        raw_sessions = find_sessions(project_root=project_root_str, vendor=active_vendor, inspect=True)
        for s in raw_sessions:
            sid = s.get("session_id")
            title = generate_standard_title(s)
            is_main = (sid == main_thread_id) if main_thread_id else False
            discovered_sessions.append({
                "session_id": sid,
                "vendor": s.get("vendor", active_vendor),
                "title": title,
                "is_main": is_main,
                "created_at": s.get("created_at"),
                "last_active_at": s.get("last_active_at"),
                "log_path": s.get("log_path"),
                "recent_prompts": s.get("recent_prompts", [])
            })

    # Prepare sessions.json content
    sessions_content = {
        "schema_version": 1,
        "main_thread_id": main_thread_id,
        "sessions": discovered_sessions if discovered_sessions else existing_sessions_data.get("sessions", []),
        "modules": existing_sessions_data.get("modules", {})
    }

    # Write sessions.json
    with open(sessions_file, "w", encoding="utf-8") as f:
        json.dump(sessions_content, f, ensure_ascii=False, indent=2)

    # Initialize other runtime files
    files_to_create = {
        "todo.json": {
            "schema_version": 2,
            "items": []
        },
        "topics.json": {
            "schema_version": 1,
            "ignored_memory_docs": [],
            "topics": []
        },
        "policy.json": {
            "schema_version": 1,
            "active_vendor": active_vendor,
            "lease_minutes": 25,
            "allow_remote_push": False,
            "model_profiles": {
                "simple": "gpt-5.6-luna",
                "standard": "gpt-5.6-terra",
                "complex": "gpt-5.6-sol"
            }
        }
    }

    for fname, content in files_to_create.items():
        fpath = root / fname
        if not fpath.is_file():
            with open(fpath, "w", encoding="utf-8") as f:
                json.dump(content, f, ensure_ascii=False, indent=2)

    journal = root / "run-journal.jsonl"
    if not journal.is_file():
        journal.touch()

    return {
        "action": "INITIALIZED",
        "root": str(root).replace("\\", "/"),
        "main_thread_id": main_thread_id,
        "active_vendor": active_vendor,
        "discovered_sessions_count": len(discovered_sessions)
    }


def main():
    parser = argparse.ArgumentParser(description="Initialize Task Loop State & Discover Sessions")
    parser.add_argument("--root", default=".agents/task-loop", help="State root directory")
    parser.add_argument("--project-root", default=".", help="Project root directory")
    parser.add_argument("--main-thread-id", default=None, help="Main thread ID")
    parser.add_argument("--clear-main", action="store_true", help="Clear main thread ID to trigger selection")
    parser.add_argument("--vendor", default="antigravity", help="Active vendor")
    parser.add_argument("--no-scan", action="store_true", help="Disable auto scanning of existing sessions")
    args = parser.parse_args()

    result = init_state(
        root_path_str=args.root,
        project_root_str=args.project_root,
        main_thread_id=args.main_thread_id,
        active_vendor=args.vendor,
        scan_sessions=not args.no_scan,
        clear_main=args.clear_main
    )
    print(json.dumps(result, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()

File: scripts/initialize_task_loop_state.js
#!/usr/bin/env node
/**
 * Initialize Task Loop State (Node.js)
 * Initializes project task-loop state files (.agents/task-loop/) with UTF-8 encoding.
 * Supports auto-scanning project sessions, standardized title generation ([具体专题] 功能点1 & 功能点2),
 * and managing main_thread_id selection.
 */

const fs = require('fs');
const path = require('path');
const { findSessions } = require('./find_project_sessions');

function generateStandardTitle(session) {
  const rawTitle = (session.title || '').replace(/\[[^\]]+\]\([^\)]+\)/g, '').trim();
  const titleLower = rawTitle.toLowerCase();
  const summaryLower = (session.summary || '').toLowerCase();
  const promptsLower = (session.recent_prompts || []).join(' ').toLowerCase();
  const touchedLower = (session.recent_touched_files || []).join(' ').toLowerCase();
  const fullText = `${titleLower} ${summaryLower} ${promptsLower} ${touchedLower}`;

  let category = '';
  let func1 = '';
  let func2 = '';

  if (fullText.includes('topic: hook') || fullText.includes('钩子专题') || (titleLower.includes('hook') && !titleLower.includes('main'))) {
    category = '钩子专题';
    func1 = '生命周期';
    func2 = '安全门禁';
  } else if (fullText.includes('topic: subagent') || fullText.includes('子代理专题') || (titleLower.includes('subagent') && !titleLower.includes('main'))) {
    category = '子代理专题';
    func1 = '动态模板';
    func2 = '编排治理';
  } else if (fullText.includes('topic: session_control') || fullText.includes('会话专题') || (titleLower.includes('session') && !titleLower.includes('main'))) {
    category = '会话控制专题';
    func1 = '跨厂商内省';
    func2 = '会话管理';
  } else if (fullText.includes('topic: memory') || fullText.includes('记忆专题') || titleLower.includes('memory')) {
    category = '受控记忆专题';
    func1 = '文档维护';
    func2 = '经验沉淀';
  } else if (fullText.includes('code review') || fullText.includes('代码审查') || fullText.includes('高级代码审查员') || fullText.includes('走查')) {
    category = '代码审查专题';
    func1 = '质量走查';
    func2 = '门禁核验';
  } else if (fullText.includes('debug') || fullText.includes('排障') || fullText.includes('卡顿') || fullText.includes('故障') || fullText.includes('报错')) {
    category = '排障诊断专题';
    func1 = '缺陷定位';
    func2 = '故障分析';
  } else if (fullText.includes('claude code') || fullText.includes('claude sdk') || fullText.includes('claude')) {
    category = 'Claude协同专题';
    func1 = 'SDK适配';
    func2 = '跨平台支持';
  } else if (fullText.includes('codex') || fullText.includes('openai')) {
    category = 'Codex协同专题';
    func1 = '跨端同步';
    func2 = '会话管理';
  } else if (fullText.includes('plugin') || fullText.includes('插件')) {
    category = 'Plugin规范专题';
    func1 = '接口定义';
    func2 = '插件集成';
  } else if (fullText.includes('dispatch') || fullText.includes('调度') || fullText.includes('task-loop') || fullText.includes('task_loop')) {
    category = '任务循环调度专题';
    func1 = '任务分发';
    func2 = '状态机管理';
  } else if (session.is_main || titleLower.includes('main') || fullText.includes('治理中枢') || fullText.includes('开发仓库')) {
    category = '主会话';
    func1 = '任务编排';
    func2 = '治理中枢';
  } else {
    const cleanText = rawTitle.replace(/[^\w\s\u4e00-\u9fa5]/g, ' ');
    const stopWords = new Set(['请你', '一个', '当前', '这个', '作为', '可以', '需要', '进行', '如何', '为什么', '是否', '实现', '相关', '检查', '项目']);
    const words = cleanText.split(/\s+/).filter(w => w.length >= 2 && !stopWords.has(w));
    const kw1 = words[0] || '核心业务';
    const kw2 = words[1] || '功能实现';
    category = `${kw1}专题`;
    func1 = kw1;
    func2 = kw2;
  }

  return `[${category}] ${func1} & ${func2}`;
}

function initState({
  rootPathStr = '.agents/task-loop',
  projectRootStr = '.',
  mainThreadId = null,
  activeVendor = 'antigravity',
  scanSessions = true,
  clearMain = false
} = {}) {
  const root = path.resolve(rootPathStr);
  if (!fs.existsSync(root)) {
    fs.mkdirSync(root, { recursive: true });
  }

  const sessionsFile = path.join(root, 'sessions.json');
  let existingSessionsData = {};
  if (fs.existsSync(sessionsFile)) {
    try {
      existingSessionsData = JSON.parse(fs.readFileSync(sessionsFile, 'utf8'));
    } catch (e) {}
  }

  if (clearMain) {
    mainThreadId = null;
  } else if (mainThreadId === null && existingSessionsData.main_thread_id) {
    mainThreadId = existingSessionsData.main_thread_id;
  }

  let discoveredSessions = [];
  if (scanSessions) {
    const rawSessions = findSessions(projectRootStr, activeVendor, true);
    for (const s of rawSessions) {
      const sid = s.session_id;
      const title = generateStandardTitle(s);
      const isMain = mainThreadId ? sid === mainThreadId : false;
      discoveredSessions.push({
        session_id: sid,
        vendor: s.vendor || activeVendor,
        title: title,
        is_main: isMain,
        created_at: s.created_at,
        last_active_at: s.last_active_at,
        log_path: s.log_path,
        recent_prompts: s.recent_prompts || []
      });
    }
  }

  const sessionsContent = {
    schema_version: 1,
    main_thread_id: mainThreadId,
    sessions: discoveredSessions.length > 0 ? discoveredSessions : (existingSessionsData.sessions || []),
    modules: existingSessionsData.modules || {}
  };

  fs.writeFileSync(sessionsFile, JSON.stringify(sessionsContent, null, 2), 'utf8');

  const filesToCreate = {
    'todo.json': {
      schema_version: 2,
      items: []
    },
    'topics.json': {
      schema_version: 1,
      ignored_memory_docs: [],
      topics: []
    },
    'policy.json': {
      schema_version: 1,
      active_vendor: activeVendor,
      lease_minutes: 25,
      allow_remote_push: false,
      model_profiles: {
        simple: 'gpt-5.6-luna',
        standard: 'gpt-5.6-terra',
        complex: 'gpt-5.6-sol'
      }
    }
  };

  for (const [fname, content] of Object.entries(filesToCreate)) {
    const fpath = path.join(root, fname);
    if (!fs.existsSync(fpath)) {
      fs.writeFileSync(fpath, JSON.stringify(content, null, 2), 'utf8');
    }
  }

  const journal = path.join(root, 'run-journal.jsonl');
  if (!fs.existsSync(journal)) {
    fs.closeSync(fs.openSync(journal, 'w'));
  }

  return {
    action: 'INITIALIZED',
    root: root.replace(/\\/g, '/'),
    main_thread_id: mainThreadId,
    active_vendor: activeVendor,
    discovered_sessions_count: discoveredSessions.length
  };
}

function main() {
  const args = process.argv.slice(2);
  let root = '.agents/task-loop';
  let projectRoot = '.';
  let mainThreadId = null;
  let clearMain = false;
  let vendor = 'antigravity';
  let noScan = false;
  let surveyMode = args.includes('--survey') || args.includes('--dry-run') || args.includes('-d');

  for (let i = 0; i < args.length; i++) {
    if (args[i] === '--root' && args[i + 1]) {
      root = args[++i];
    } else if (args[i] === '--project-root' && args[i + 1]) {
      projectRoot = args[++i];
    } else if (args[i] === '--main-thread-id' && args[i + 1]) {
      mainThreadId = args[++i];
    } else if (args[i] === '--clear-main') {
      clearMain = true;
    } else if (args[i] === '--vendor' && args[i + 1]) {
      vendor = args[++i];
    } else if (args[i] === '--no-scan') {
      noScan = true;
    }
  }

  if (surveyMode) {
    console.log('================================================================================');
    console.log(' [task-loop] 项目初始化会话调查与专题映射建议 (Dry-Run Preview)');
    console.log('================================================================================');
    const rawSessions = findSessions(projectRoot, vendor, true);
    console.log(`工作区根路径: ${path.resolve(projectRoot)}`);
    console.log(`已发现历史会话总数: ${rawSessions.length}`);
    console.log(`状态机落盘目标: ${path.resolve(root, 'sessions.json')}`);
    console.log('--------------------------------------------------------------------------------');
    console.log('建议专题映射清单 (Topic Mapping Recommendations):');
    rawSessions.forEach((s, idx) => {
      const title = generateStandardTitle(s);
      console.log(`\n[${idx + 1}] 会话 ID: ${s.session_id}`);
      console.log(`    原标题/厂商: ${s.title || '(无标题)'} (${s.vendor})`);
      console.log(`    建议标准化专题名: ${title}`);
      console.log(`    最近交互提示词: ${(s.recent_prompts && s.recent_prompts[0]) ? s.recent_prompts[0].slice(0, 60) + '...' : '(无)'}`);
    });
    console.log('\n================================================================================');
    console.log('【预览完成】确认上述映射后，去掉 --dry-run / --survey 参数即可实际落盘写入。');
    console.log(`【兜底修改提示】写入后若有命名/ID微调需求，可直接打开编辑: ${path.resolve(root, 'sessions.json')}`);
    console.log('================================================================================\n');
    return;
  }

  const result = initState({
    rootPathStr: root,
    projectRootStr: projectRoot,
    mainThreadId: mainThreadId,
    activeVendor: vendor,
    scanSessions: !noScan,
    clearMain: clearMain
  });

  console.log(JSON.stringify(result, null, 2));
}

if (require.main === module) {
  main();
}

module.exports = { initState, generateStandardTitle };
